File: frontend/src/api/userSettings.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { api } from './client';

export type UserSettingsKey = 'display' | 'nutrition_calendar' | 'cumulative_fluid_chart' | 'developer_mode' | 'med_nudge';

/** @deprecated Use UserSettingsKey */
export type WidgetSettingsKey = Exclude<UserSettingsKey, 'display'>;

export type WeekStart = 'sunday' | 'monday';
export type TimeFormat = 'h24' | 'h12';
export type DateFormat = 'dmy' | 'mmm_dd_yyyy';

/** Per-user display preferences (time/date formatting, page toggles). */
export interface UserDisplaySettings {
  time_format: TimeFormat;
  date_format: DateFormat;
  show_water_card: boolean;
}

export interface NutritionCalendarSettings {
  week_start: WeekStart;
  show_wet_food: boolean;
  show_liquids: boolean;
  show_water: boolean;
  show_dry_food: boolean;
  show_record_count: boolean;
  show_total_fluid: boolean;
}

export interface CumulativeFluidChartSettings {
  show_current_liquids: boolean;
  show_current_food_fluid: boolean;
  show_current_total: boolean;
  show_best_day_liquids: boolean;
  show_best_day_food_fluid: boolean;
  show_best_day_total: boolean;
  show_schedule: boolean;
  show_now_bar: boolean;
}

/** Per-user developer tooling (curl snippets, etc.). */
export interface DeveloperModeSettings {
  enabled: boolean;
}

export interface NudgeSlot {
  enabled: boolean;
  /** Hour of day 0–23 by which the dose should have been logged. */
  deadline_hour: number;
}

export interface PetNudgeSchedule {
  morning: NudgeSlot;
  midday: NudgeSlot;
  evening: NudgeSlot;
}

/** Per-user medication nudge settings, keyed by pet_id. */
export interface MedNudgeSettings {
  pets: Record<string, PetNudgeSchedule>;
}

export type UserSettingsMap = {
  display: UserDisplaySettings;
  nutrition_calendar: NutritionCalendarSettings;
  cumulative_fluid_chart: CumulativeFluidChartSettings;
  developer_mode: DeveloperModeSettings;
  med_nudge: MedNudgeSettings;
};

export const DEFAULT_USER_DISPLAY_SETTINGS: UserDisplaySettings = {
  time_format: 'h24',
  date_format: 'dmy',
  show_water_card: true,
};

export const DEFAULT_NUTRITION_CALENDAR_SETTINGS: NutritionCalendarSettings = {
  week_start: 'sunday',
  show_wet_food: true,
  show_liquids: true,
  show_water: true,
  show_dry_food: true,
  show_record_count: true,
  show_total_fluid: true,
};

export const DEFAULT_CUMULATIVE_FLUID_CHART_SETTINGS: CumulativeFluidChartSettings = {
  show_current_liquids: true,
  show_current_food_fluid: false,
  show_current_total: false,
  show_best_day_liquids: true,
  show_best_day_food_fluid: false,
  show_best_day_total: false,
  show_schedule: true,
  show_now_bar: true,
};

export const DEFAULT_DEVELOPER_MODE_SETTINGS: DeveloperModeSettings = {
  enabled: false,
};

export const DEFAULT_NUDGE_SLOT: NudgeSlot = { enabled: false, deadline_hour: 9 };

export const DEFAULT_PET_NUDGE_SCHEDULE: PetNudgeSchedule = {
  morning: { enabled: false, deadline_hour: 9 },
  midday: { enabled: false, deadline_hour: 13 },
  evening: { enabled: false, deadline_hour: 20 },
};

export const DEFAULT_MED_NUDGE_SETTINGS: MedNudgeSettings = { pets: {} };

export const USER_SETTINGS_DEFAULTS: UserSettingsMap = {
  display: DEFAULT_USER_DISPLAY_SETTINGS,
  nutrition_calendar: DEFAULT_NUTRITION_CALENDAR_SETTINGS,
  cumulative_fluid_chart: DEFAULT_CUMULATIVE_FLUID_CHART_SETTINGS,
  developer_mode: DEFAULT_DEVELOPER_MODE_SETTINGS,
  med_nudge: DEFAULT_MED_NUDGE_SETTINGS,
};

export const userSettingsApi = {
  get: <K extends UserSettingsKey>(key: K) =>
    api.get<UserSettingsMap[K]>(`/me/settings/${key}`),

  update: <K extends UserSettingsKey>(key: K, body: Partial<UserSettingsMap[K]>) =>
    api.post<UserSettingsMap[K]>(`/me/settings/${key}`, body),
};

export function userSettingsQueryKey(key: UserSettingsKey) {
  return ['user-settings', key] as const;
}

export function useUserSettings<K extends UserSettingsKey>(key: K) {
  const queryClient = useQueryClient();
  const queryKey = userSettingsQueryKey(key);

  const query = useQuery({
    queryKey,
    queryFn: () => userSettingsApi.get(key),
    staleTime: Infinity,
  });

  const mutation = useMutation({
    mutationFn: (patch: Partial<UserSettingsMap[K]>) => userSettingsApi.update(key, patch),
    onSuccess: (updated) => {
      queryClient.setQueryData(queryKey, updated);
    },
  });

  const settings = query.data ?? USER_SETTINGS_DEFAULTS[key];

  function update(patch: Partial<UserSettingsMap[K]>) {
    mutation.mutate(patch);
  }

  return { settings, update, isLoading: query.isLoading, isSaving: mutation.isPending, error: mutation.error ?? query.error };
}

/** Widget-only alias — calendar and chart gear controls. */
export function useUserWidgetSettings<K extends WidgetSettingsKey>(key: K) {
  return useUserSettings(key);
}

export function widgetSettingsQueryKey(key: WidgetSettingsKey) {
  return userSettingsQueryKey(key);
}

File: src/domain/user_settings.rs
use serde::{Deserialize, Serialize};

use super::settings::{DateFormat, TimeFormat, WeekStart};

pub const DISPLAY_KEY: &str = "display";
pub const NUTRITION_CALENDAR_KEY: &str = "nutrition_calendar";
pub const CUMULATIVE_FLUID_CHART_KEY: &str = "cumulative_fluid_chart";
pub const DEVELOPER_MODE_KEY: &str = "developer_mode";

/// Per-user display preferences (global UI formatting, page toggles).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserDisplaySettings {
    #[serde(default)]
    pub time_format: TimeFormat,
    #[serde(default)]
    pub date_format: DateFormat,
    #[serde(default = "default_true")]
    pub show_water_card: bool,
}

impl Default for UserDisplaySettings {
    fn default() -> Self {
        UserDisplaySettings {
            time_format: TimeFormat::default(),
            date_format: DateFormat::default(),
            show_water_card: true,
        }
    }
}

#[derive(Debug, Deserialize)]
pub struct UpdateUserDisplaySettings {
    pub time_format: Option<TimeFormat>,
    pub date_format: Option<DateFormat>,
    pub show_water_card: Option<bool>,
}

impl UpdateUserDisplaySettings {
    pub fn apply(self, existing: UserDisplaySettings) -> UserDisplaySettings {
        UserDisplaySettings {
            time_format: self.time_format.unwrap_or(existing.time_format),
            date_format: self.date_format.unwrap_or(existing.date_format),
            show_water_card: self.show_water_card.unwrap_or(existing.show_water_card),
        }
    }
}

/// Per-user nutrition journal calendar widget preferences.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NutritionCalendarSettings {
    #[serde(default)]
    pub week_start: WeekStart,
    #[serde(default = "default_true")]
    pub show_wet_food: bool,
    #[serde(default = "default_true")]
    pub show_liquids: bool,
    #[serde(default = "default_true")]
    pub show_water: bool,
    #[serde(default = "default_true")]
    pub show_dry_food: bool,
    #[serde(default = "default_true")]
    pub show_record_count: bool,
    #[serde(default = "default_true")]
    pub show_total_fluid: bool,
}

impl Default for NutritionCalendarSettings {
    fn default() -> Self {
        NutritionCalendarSettings {
            week_start: WeekStart::default(),
            show_wet_food: true,
            show_liquids: true,
            show_water: true,
            show_dry_food: true,
            show_record_count: true,
            show_total_fluid: true,
        }
    }
}

#[derive(Debug, Deserialize)]
pub struct UpdateNutritionCalendarSettings {
    pub week_start: Option<WeekStart>,
    pub show_wet_food: Option<bool>,
    pub show_liquids: Option<bool>,
    pub show_water: Option<bool>,
    pub show_dry_food: Option<bool>,
    pub show_record_count: Option<bool>,
    pub show_total_fluid: Option<bool>,
}

impl UpdateNutritionCalendarSettings {
    pub fn apply(self, existing: NutritionCalendarSettings) -> NutritionCalendarSettings {
        NutritionCalendarSettings {
            week_start: self.week_start.unwrap_or(existing.week_start),
            show_wet_food: self.show_wet_food.unwrap_or(existing.show_wet_food),
            show_liquids: self.show_liquids.unwrap_or(existing.show_liquids),
            show_water: self.show_water.unwrap_or(existing.show_water),
            show_dry_food: self.show_dry_food.unwrap_or(existing.show_dry_food),
            show_record_count: self.show_record_count.unwrap_or(existing.show_record_count),
            show_total_fluid: self.show_total_fluid.unwrap_or(existing.show_total_fluid),
        }
    }
}

/// Per-user cumulative fluid chart widget preferences.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CumulativeFluidChartSettings {
    #[serde(default = "default_true")]
    pub show_current_liquids: bool,
    #[serde(default = "default_false")]
    pub show_current_food_fluid: bool,
    #[serde(default = "default_false")]
    pub show_current_total: bool,
    #[serde(default = "default_true")]
    pub show_best_day_liquids: bool,
    #[serde(default = "default_false")]
    pub show_best_day_food_fluid: bool,
    #[serde(default = "default_false")]
    pub show_best_day_total: bool,
    #[serde(default = "default_true")]
    pub show_schedule: bool,
    #[serde(default = "default_true")]
    pub show_now_bar: bool,
}

impl Default for CumulativeFluidChartSettings {
    fn default() -> Self {
        CumulativeFluidChartSettings {
            show_current_liquids: true,
            show_current_food_fluid: false,
            show_current_total: false,
            show_best_day_liquids: true,
            show_best_day_food_fluid: false,
            show_best_day_total: false,
            show_schedule: true,
            show_now_bar: true,
        }
    }
}

#[derive(Debug, Deserialize)]
pub struct UpdateCumulativeFluidChartSettings {
    pub show_current_liquids: Option<bool>,
    pub show_current_food_fluid: Option<bool>,
    pub show_current_total: Option<bool>,
    pub show_best_day_liquids: Option<bool>,
    pub show_best_day_food_fluid: Option<bool>,
    pub show_best_day_total: Option<bool>,
    pub show_schedule: Option<bool>,
    pub show_now_bar: Option<bool>,
}

impl UpdateCumulativeFluidChartSettings {
    pub fn apply(self, existing: CumulativeFluidChartSettings) -> CumulativeFluidChartSettings {
        CumulativeFluidChartSettings {
            show_current_liquids: self
                .show_current_liquids
                .unwrap_or(existing.show_current_liquids),
            show_current_food_fluid: self
                .show_current_food_fluid
                .unwrap_or(existing.show_current_food_fluid),
            show_current_total: self
                .show_current_total
                .unwrap_or(existing.show_current_total),
            show_best_day_liquids: self
                .show_best_day_liquids
                .unwrap_or(existing.show_best_day_liquids),
            show_best_day_food_fluid: self
                .show_best_day_food_fluid
                .unwrap_or(existing.show_best_day_food_fluid),
            show_best_day_total: self
                .show_best_day_total
                .unwrap_or(existing.show_best_day_total),
            show_schedule: self.show_schedule.unwrap_or(existing.show_schedule),
            show_now_bar: self.show_now_bar.unwrap_or(existing.show_now_bar),
        }
    }
}

fn default_true() -> bool {
    true
}

fn default_false() -> bool {
    false
}

/// Per-user developer tooling (curl snippets, etc.).
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct DeveloperModeSettings {
    #[serde(default)]
    pub enabled: bool,
}

#[derive(Debug, Deserialize)]
pub struct UpdateDeveloperModeSettings {
    pub enabled: Option<bool>,
}

impl UpdateDeveloperModeSettings {
    pub fn apply(self, existing: DeveloperModeSettings) -> DeveloperModeSettings {
        DeveloperModeSettings {
            enabled: self.enabled.unwrap_or(existing.enabled),
        }
    }
}

pub fn is_known_user_settings_key(key: &str) -> bool {
    matches!(
        key,
        DISPLAY_KEY | NUTRITION_CALENDAR_KEY | CUMULATIVE_FLUID_CHART_KEY | DEVELOPER_MODE_KEY
    )
}
